File: client/SkorpioFrigo/Protocol.cs
using System;
using System.Text;

namespace SkorpioFrigo
{
    /// <summary>
    /// Protocole de ligne, miroir exact de frigo/protocol.py cote serveur.
    /// Toute modification ici doit etre repercutee la-bas, et inversement.
    /// </summary>
    internal static class Protocol
    {
        public const string Version = "1";
        public const char Sep = '|';
        public const string LineEnd = "\r\n";
        public const char ChecksumMark = '*';

        /// <summary>XOR de tous les octets, sur deux chiffres hexadecimaux.</summary>
        public static string Checksum(string payload)
        {
            byte[] bytes = Encoding.UTF8.GetBytes(payload);
            int acc = 0;
            for (int i = 0; i < bytes.Length; i++)
                acc ^= bytes[i];
            return acc.ToString("X2");
        }

        /// <summary>Retire separateurs et caracteres de controle d'un champ.</summary>
        public static string Sanitize(string value)
        {
            if (value == null)
                return "";
            StringBuilder sb = new StringBuilder(value.Length);
            for (int i = 0; i < value.Length; i++)
            {
                char c = value[i];
                if (c == Sep || c == ChecksumMark || c < ' ')
                    sb.Append(' ');
                else
                    sb.Append(c);
            }
            return sb.ToString().Trim();
        }

        /// <summary>Construit une ligne complete, checksum incluse, sans CRLF.</summary>
        public static string Encode(params string[] parts)
        {
            StringBuilder sb = new StringBuilder();
            for (int i = 0; i < parts.Length; i++)
            {
                if (i > 0)
                    sb.Append(Sep);
                sb.Append(i == 0 ? Sanitize(parts[i]).ToUpper() : Sanitize(parts[i]));
            }
            string payload = sb.ToString();
            return payload + Sep + ChecksumMark + Checksum(payload);
        }

        /// <summary>
        /// Decoupe une ligne recue et verifie la checksum si elle est presente.
        /// Retourne les champs, le verbe en position 0.
        /// </summary>
        public static string[] Decode(string line)
        {
            if (line == null)
                throw new FormatException("ligne vide");
            string text = line.Trim();
            if (text.Length == 0)
                throw new FormatException("ligne vide");

            string[] raw = text.Split(Sep);
            int count = raw.Length;
            if (count > 1 && raw[count - 1].Length > 0 && raw[count - 1][0] == ChecksumMark)
            {
                string received = raw[count - 1].Substring(1).ToUpper();
                count--;
                StringBuilder sb = new StringBuilder();
                for (int i = 0; i < count; i++)
                {
                    if (i > 0)
                        sb.Append(Sep);
                    sb.Append(raw[i]);
                }
                string expected = Checksum(sb.ToString());
                if (received != expected)
                    throw new FormatException("checksum " + received + " attendue " + expected);
            }

            string[] fields = new string[count];
            for (int i = 0; i < count; i++)
                fields[i] = raw[i].Trim();
            fields[0] = fields[0].ToUpper();
            return fields;
        }

        /// <summary>Champ a l'indice demande, chaine vide si absent.</summary>
        public static string Field(string[] fields, int index)
        {
            return (fields != null && index < fields.Length) ? fields[index] : "";
        }

        /// <summary>Horodatage ISO local, a la seconde (le terminal n'a pas de fuseau).</summary>
        public static string NowIso()
        {
            return DateTime.Now.ToString("yyyy-MM-dd'T'HH:mm:ss");
        }
    }
}

File: frigo/protocol.py
"""Protocole de ligne entre le client Skorpio et le serveur PC.

Un seul format sert les trois transports (USB/serie, TCP, fichier sur carte
memoire) : des lignes ASCII terminees par CRLF, champs separes par '|', avec
une somme de controle XOR facultative en fin de ligne.

    SCAN|12|3017620422003|2026-10-25|1|frigo|2026-09-07T18:04:11|*4F
    ^    ^  ^             ^          ^ ^     ^                    ^
    |    |  code-barres   peremption | lieu  horodatage terminal  checksum
    |    numero de sequence          quantite
    verbe

Le couple (device, seq) rend l'ingestion idempotente : reposer le terminal sur
son socle et renvoyer tout le tampon ne cree jamais de doublon.

Terminal -> serveur
    HELLO|<device>|<proto>|<version_client>
    SCAN|<seq>|<code>|<peremption>|<qte>|<lieu>|<horodatage>|<note>
    LOOK|<code>                      demande de libelle produit
    BYE|<nb_scans_envoyes>
    PING

Serveur -> terminal
    READY|<proto>|<version_serveur>|<device>
    OK|<seq>|<libelle>               scan enregistre
    DUP|<seq>                        deja connu, ignore
    ERR|<seq>|<motif>
    INFO|<code>|<libelle>|<marque>
    PONG
"""

from __future__ import annotations

import datetime as dt
from dataclasses import dataclass, field

__all__ = [
    "PROTO_VERSION",
    "FIELD_SEP",
    "LINE_END",
    "Message",
    "ProtocolError",
    "encode",
    "decode",
    "checksum",
    "sanitize",
    "now_iso",
]

PROTO_VERSION = "1"
FIELD_SEP = "|"
LINE_END = "\r\n"
CHECKSUM_MARK = "*"
MAX_LINE = 1024


class ProtocolError(ValueError):
    """Ligne inexploitable (verbe inconnu, checksum fausse, ligne tronquee)."""


@dataclass(slots=True)
class Message:
    verb: str
    args: list[str] = field(default_factory=list)

    def arg(self, index: int, default: str = "") -> str:
        try:
            return self.args[index]
        except IndexError:
            return default

    def __str__(self) -> str:
        return encode(self.verb, *self.args, with_checksum=False)


def checksum(payload: str) -> str:
    """XOR de tous les octets de la charge utile, sur deux chiffres hex."""
    acc = 0
    for byte in payload.encode("utf-8", "replace"):
        acc ^= byte
    return f"{acc:02X}"


def sanitize(value: object) -> str:
    """Rend une valeur transportable : pas de separateur, pas de caractere de controle."""
    if value is None:
        return ""
    text = str(value)
    out = []
    for ch in text:
        if ch == FIELD_SEP or ch == CHECKSUM_MARK or ord(ch) < 0x20:
            out.append(" ")
        else:
            out.append(ch)
    return "".join(out).strip()


def encode(verb: str, *args: object, with_checksum: bool = True) -> str:
    """Construit une ligne (sans le CRLF final)."""
    payload = FIELD_SEP.join([verb.upper()] + [sanitize(a) for a in args])
    if not with_checksum:
        return payload
    return f"{payload}{FIELD_SEP}{CHECKSUM_MARK}{checksum(payload)}"


def decode(line: str, *, strict_checksum: bool = False) -> Message:
    """Analyse une ligne recue. La somme de controle est verifiee si presente."""
    if line is None:
        raise ProtocolError("ligne vide")
    text = line.strip("\r\n").strip()
    if not text:
        raise ProtocolError("ligne vide")
    if len(text) > MAX_LINE:
        raise ProtocolError(f"ligne trop longue ({len(text)} caracteres)")

    parts = text.split(FIELD_SEP)
    if parts and parts[-1].startswith(CHECKSUM_MARK):
        received = parts.pop()[len(CHECKSUM_MARK):].upper()
        payload = FIELD_SEP.join(parts)
        expected = checksum(payload)
        if received != expected:
            raise ProtocolError(f"checksum {received} attendue {expected}")
    elif strict_checksum:
        raise ProtocolError("checksum absente")

    verb = parts[0].strip().upper()
    if not verb.isalpha():
        raise ProtocolError(f"verbe invalide : {parts[0]!r}")
    return Message(verb=verb, args=[p.strip() for p in parts[1:]])


def now_iso() -> str:
    """Horodatage local a la seconde, sans fuseau (le terminal n'en a pas)."""
    return dt.datetime.now().replace(microsecond=0).isoformat()
